test/functional/test_framework/test_node.py

```diff
@@ -98,14 +98,23 @@ def __init__(self, i, datadir, *, chain, rpchost, timewait, bitcoind, bitcoin_cl
             "-uacomment=testnode%d" % i,
         ]
         if use_valgrind:
-            default_suppressions_file = os.path.join(
-                os.path.dirname(os.path.realpath(__file__)),
-                "..", "..", "..", "contrib", "valgrind.supp")
-            suppressions_file = os.getenv("VALGRIND_SUPPRESSIONS_FILE",
-                                          default_suppressions_file)
-            self.args = ["valgrind", "--suppressions={}".format(suppressions_file),
-                         "--gen-suppressions=all", "--exit-on-first-error=yes",
-                         "--error-exitcode=1", "--quiet"] + self.args
+            valgrind_version = self.get_valgrind_version()
+            exit_on_first_error_flag = ""
+            if valgrind_version and valgrind_version >= MIN_VALGRIND_VERSION_EARLY_EXIT:
+                exit_on_first_error_flag = "--exit-on-first-error=yes"
+                default_suppressions_file = os.path.join(
+                    os.path.dirname(os.path.realpath(__file__)),
+                    "..", "..", "..", "contrib", "valgrind.supp")
+                suppressions_file = os.getenv("VALGRIND_SUPPRESSIONS_FILE",
+                    default_suppressions_file)
+                self.args = ["valgrind",
+                             "--suppressions={}".format(suppressions_file),
+                             "--gen-suppressions=all",
+                             exit_on_first_error_flag,
+                             "--error-exitcode=1",
+                             "--quiet"] + self.args
+            elif valgrind_version == None:
```

Comment:
I think you want to do this check right after `valgrind_version = self.get_valgrind_version()`. With the current logic the exception will be thrown also in the case where `valgrind` is present (but with an older version).

I suggest testing the PR with the following alternatives to make sure all code paths are working as expected:

1.) Without `valgrind` in `PATH`
2.) With new `valgrind` version (>= 3.14) in `PATH`
3.) With old `valgrind` version (< 3.14) in `PATH`

:)